site/docs/skylark/macros.md

```diff
@@ -7,12 +7,131 @@ title: Macros
 
 <!-- [TOC] -->
 
-## Macro creation
+## Introduction
 
 A macro is a function called from the BUILD file that can instantiate rules.
-Macros don't give additional power, they are just used for encapsulation and
-code reuse. By the end of the [loading phase](concepts.md#evaluation-model),
-macros don't exist anymore, and Bazel sees only the set of rules they created.
+Macros are mainly used for encapsulation and code reuse of existing rules 
+and other macros. By the end of the [loading phase](concepts.md#evaluation-model), 
+macros don't exist anymore, and Bazel sees only the concrete set of instantiated rules.
+
+## Usage
+
+The typical use-case for a macro is when you want to reuse a rule. 
+
+For example, we have a genrule in a BUILD file that generates a file 
+using `//test:generator` with a `some_arg` argument hardcoded in the command:
+
+```python
+genrule(
+    name = "file",
+    outs = ["file.txt"],
+    cmd = "$(location generator) some_arg > $@",
+    tools = ["//test:generator"],
+)
+```
+
+> Tip: `$@` is a [Make variable](../be/make-variables.html#predefined_genrule_variables) 
+> that refers to the execution-time locations of the files in the `outs` attribute list. 
+> It is equivalent to `$(locations :file.txt)`.
+
+If you want to generate more files with different arguments, you may want to
+extract this code to a macro function. Let's call the macro `file_generator`, which
+has `name` and `arg` parameters. Replace the genrule with the following:
+
+```python
+load("//path:generator.bzl", "file_generator")
+
+file_generator(
+    name = "file",
+    arg = "some_arg",
+)
+
+file_generator(
+    name = "file-two",
+    arg = "some_arg_two",
+)
+
+file_generator(
+    name = "file-three",
+    arg = "some_arg_three",
+)
+```
+
+Here, we are loading the `file_generator` symbol from a `.bzl` file located in the 
+`//path` package. By putting macro function definitions in a separate `.bzl` file, 
+we can keep our BUILD files clean and declarative,  The `.bzl` file can be loaded 
+from any package in the workspace, as long as `visibility` conditions are met.
+
+> Tip: To make a `.bzl` file visible from a package, add [`exports_files(["<filename>.bzl"])`](../be/functions.html#exports_files) into the `BUILD` file in the same directory as the `.bzl` file.
```

Comment:
exports_files is not recommended

Reply:
Are you recommending dropping the paragraph or replacing it with a pointer to bzl_library?

Reply:
Not recommended at all? Also, it seems that we don't need it -- just an empty BUILD file will do.

Reply:
Thanks. Removed.